Updated LDAS Tools Suite of tools

emaros opened the pull request:
Added port:pkgconfig as build dependency; Do not build on anything less than Lion

lpsinger commented:
What is not supported on pre-Lion releases?

emaros commented:
Pre-Lion is where the PPC systems are still active and the buildbot reported errors. Since I don't have a system that old, it would be difficult to iterate over changes to make it work. I believe some of the issues are related to C++ standard conformance.

mojca commented:
Please provide a few links to buildbot errors at least. Sometimes fixing a problem might be as easy as blacklisting a few old compilers.

emaros commented:
Here is an example:

https://build.macports.org/builders/ports-10.6_i386_legacy-builder/builds/9735/steps/install-dependencies/logs/stdio

lpsinger commented:
> Pre-Lion is where the PPC systems are still active and the buildbot reported errors. Since I don't have a system that old, it would be difficult to iterate over changes to make it work. I believe some of the issues are related to C++ standard conformance.

If it's a PPC issue, then you should use `supported_archs` (see https://guide.macports.org/#reference.keywords).

lpsinger commented:
This doesn't look like a PPC issue, though. Does your code require a bleeding-edge C++ standard library?

emaros commented:
Not bleeding edge, but at least TR1 and I am trying to move to 2011 standard

lpsinger commented:
Maybe try building with `-std=c++11`?

mojca commented:
The failure is [in this build](https://build.macports.org/builders/ports-10.6_i386_legacy-builder/builds/9745) for the port `ldas-tools-al`:
~~~
Task.cc:66:54: error: no member named 'default_delete' in namespace 'std'
      name.reset( new char[Name.length( ) + 1], std::default_delete<char[]>() );
                                                ~~~~~^
Task.cc:66:73: error: expected '(' for function-style cast or type construction
      name.reset( new char[Name.length( ) + 1], std::default_delete<char[]>() );
                                                                    ~~~~^
2 errors generated.
~~~
This looks like the software might require some features of C++11 to build rather than something that is inherently impossible to compile on 10.6 and below.

This doesn't look like the right patch to me.

emaros commented:
Maybe the correct solution is to blacklist any version of XCode prior to the default that comes with 10.7? Where could I find this information and what is the best way to test the fix?

emaros commented:
I have only addressed the base library to determine if this is sufficient. If it works for ldas-tools-al, I will replicate for the other packages.

mojca commented:
After I saw the error and noticed that it was built with clang-3.4, I doubt this is just outdated compiler, but perhaps too old stdlib. If you switch to 3.7, it might use libc++ and then work just because of that? This is pure speculation though.

emaros commented:
Why doesn't the line:

compiler.blacklist-append {clang < 500.2.75} llvm-gcc-4.2 gcc-4.2

prevent the use of clan-3.4?

lpsinger commented:
> Why doesn't the line:
> 
> compiler.blacklist-append {clang < 500.2.75} llvm-gcc-4.2 gcc-4.2
> 
> prevent the use of clan-3.4?

I don't know, but I tend to use `compiler.whitelist`/`compiler.fallback`. See [py-healpy](https://github.com/macports/macports-ports/blob/master/python/py-healpy/Portfile), for instance.

mojca commented:
I guess that `clang < 500.2.75` only prevents system clang older than Xcode 5

ryandesign commented:
Requiring a particular Xcode version is unlikely to be the correct solution. More likely, you're running into problems related to an old compiler, an old C++ library, or an old OS.

`compiler.blacklist { clang < 500.2.75 }` only blacklists the compiler known as "clang". It does not blacklist the compiler known as "macports-clang-3.4" because, as far as MacPorts base is concerned, that's a completely different and unrelated compiler.

Using `compiler.blacklist` is preferred over using `compiler.whitelist` and `compiler.fallback`.

emaros commented:
Since I need at least 500.2.75 of clang, can I add:
compiler.blacklist-append { macports-clang < 3.7 }

ryandesign commented:
No; there is no compiler called "macports-clang".

mojca commented:
Please note that using `macports-clang-3.7` alone doesn't solve the problem on 10.6. Most likely one needs `libc++` on 10.6 (or the latest gcc), while `libstdc++` and `macports-clang-3.4` still work fine on 10.7 (until the full switch to C++11). I tried compiling the port against `libc++` and that worked.

emaros commented:
The configure script tries to find the appropriate C++ libraries by trying the most "advanced" versions first. So, if the options are supported by the compiler, it should pick them up.

mojca commented:
I'm confused about something else. You opened a number of tickets on Trac ([#52799](https://trac.macports.org/ticket/52799), [#52800](https://trac.macports.org/ticket/52800), , [#52801](https://trac.macports.org/ticket/52801), [#52802](https://trac.macports.org/ticket/52802), [#52803](https://trac.macports.org/ticket/52803), [#52804](https://trac.macports.org/ticket/52804), [#52805](https://trac.macports.org/ticket/52805), [#52800](https://trac.macports.org/ticket/52800), [#52808](https://trac.macports.org/ticket/52808)). Those tickets request adding
~~~tcl
depends_build  port:autoconf port:automake port:libtool
~~~
or
~~~tcl
depends_build  port:autoconf port:automake port:libtool port:pkgconfig
~~~
while in this ticket you added (just) `pkgconfig` (as something unrelated to the choice of compiler).

* It's not clear to me which build dependencies are actually needed.
* There's no need to open so many trac tickets with "almost the same content". Next time opening a single one with all patches together (either in one file or multiple files) should be sufficient. No need to change anything now.
* Don't try to open issues on trac and PRs here for the *same* thing (I'm not saying that you did it this time), but when changes/issues are related, it's usually helpful to link from one to the other.
* I'm not sure if it's a "bug" in our infrastructure, but the email shown on those trac ticket is different from your email listed in the Portfiles.

I would suggest to fix the build dependencies first and close all those trac tickets and think about compilers and stdlibs separately.

emaros commented:
I filed those tickets under my SVN knowledge of how to request updates.
Since I had so many closely related tickets, I was encouraged to do it as a pull request.

Those tickets were being held up since I had required autoconf, automake and libtool which truly is an error. The only new build dependency should have been pkgconfig.

In addition, the buildbot started reporting build failures for ppc and I am trying to get those cleaned up so as to provide the best Portfiles for these packages that I maintain.

The tickets should probably be closed since the effort is now on getting the pull request to work properly.

Sorry if I am doing things wrong, still getting use to the new model.

mojca commented on science/ldas-tools-al/Portfile:
```diff
@@ -34,14 +35,17 @@ configure.args PYTHON=false \
-# requires clang from Xcode5 or higher to build
+# requires clang from Xcode 5.0 or higher to build
 compiler.blacklist-append {clang < 500.2.75} llvm-gcc-4.2 gcc-4.2
+compiler.blacklist-append macports-clang-3.6  macports-clang-3.5 macports-clang-3.4
```
I don't understand why these clang compilers would be blacklisted. In particular `macports-clang-3.4` works just fine on 10.7 and you could in fact get `macports-clang-3.3` with the above list.

I'm not sure, but is `macports-clang-3.7` actually on the fallback list?

emaros replied:
I trying to leverage your response from 6 hours prior.

emaros commented:
Is there a way that I can test my changes on a 10.6 box so I don't have to commit my best guesses to this branch?

emaros commented:
Would it be best if I separate the two tasks? I could create a new branch which is just the pkgconfig  change and then limit the work here to the ppc issue.

mojca commented:
Yes, that would be my suggestion as well. I would suggest that you move the part about compiler and stdlib selection to the bug tracker on Trac and put the link here because we're currently not sure what the proper fix should be (and thus it doesn't make much sense to keep using a pull request). It's best if you prepare a separate patch that includes just adding the missing dependency (it can be a single commit, you can either change this PR or add it to one of the existing Trac tickets for those ports).

ken-cunningham-webuse commented:
please don't set stdlib to libc++ in a portfile. I believe will totally mess up some systems that aren't set up that way, and do nothing to those that are.

+if {${configure.cxx} eq "macports-clang-3.7" } {
 +    configure.cxx_stdlib "libc++"
 +}
 +

ken-cunningham-webuse commented:
On 10.6, with clang-3.7 and libc++

../include/ldastoolsal/SharedPtr.hh:11:30: note: expanded from macro 'SHARED_PTR_NAMESPACE'
#define SHARED_PTR_NAMESPACE NAMESPACE_GENERAL_MEMORY
                             ^
In file included from AtExit.cc:15:
In file included from ../include/ldastoolsal/ReadWriteLock.hh:10:
In file included from ../include/ldastoolsal/unordered_map.hh:39../include/ldastoolsal/SharedPtr.hh::
In file included from /opt/local/libexec/llvm-3.7/bin/../include/c++/v1/ext/hash_map:38:203:
/opt/local/libexec/llvm-3.7/bin/../include/c++/v1/__hash_table:795:13: error: too many template arguments for class template 'auto_ptr'
    typedef unique_ptr<__node_pointer[], __bucket_list_deleter> __bucket_list;
9: error: use of undeclared identifier 'NAMESPACE_GENERAL_MEMORY'            ^                            ~~~~~~~~~~~~~~~~~~~~~~
  using SHARED_PTR_NAMESPACE::StaticPointerCast;
        ^
../include/ldastoolsal/SharedPtr.hh:11:30: note: expanded from macro 'SHARED_PTR_NAMESPACE'
#define SHARED_PTR_NAMESPACE NAMESPACE_GENERAL_MEMORY

==============

However, on 10.5 PPC with gcc6, build succeeds without any problem:

LeopardG5:~ $ port -v installed ldas-tools-al
The following ports are currently installed:
  ldas-tools-al @2.5.5_0 (active) platform='darwin 9' archs='ppc'

mojca commented:
Dependency was added in #20 / 375f4937930d7d1af70c6b5e86554263ac67d615. Please open a trac ticket to discuss the rest.

emaros commented:
I have rebased master to upstream master and then rebased this branch with local master so as to pick up the depends_build fix of pull request #20 .

I have also updated compiler.blacklist and removed setting of stdlib based on information from ken-cunningham-webuse posting.
